sources/lib/api/gcp/bigquery/_udf.py

```diff
@@ -120,19 +84,77 @@ def __init__(self, inputs, outputs, name, implementation):
       implementation: a javascript function implementing the logic.
+      support_code: additional javascript code that the function can use.
+      imports: a list of GCS URLs or files containing further support code.
     Raises:
       Exception if the name is invalid.
       """
-    self._inputs = inputs
     self._outputs = outputs
     self._name = name
     self._implementation = implementation
+    self._support_code = support_code
+    self._imports = imports
+    self._code = UDF._build_js(inputs, outputs, name, implementation, support_code)
+
+  def expand_imported_code(self):
+    """ For testing, we need to pull in any code from GCS imports. """
+    code = self._support_code if self._support_code else ''
+    if self._imports:
+      for url in self._imports:
+        try:
+          content = gcp.storage.Item.from_url(url).read_from()
+          code += '\n'
+          code += content
+        except Exception as e:
+          raise Exception('Could not read import %s' % url)
+    return code
 
   def __call__(self, data):
     if issubclass(type(data), list):
-      return FunctionEvaluation(self._implementation, data)
+      return FunctionEvaluation(self, data)
     else:
-      return FunctionCall(data, self._inputs, self._outputs, self._name, self._implementation)
+      return FunctionCall(self, UDF._build_sql(self.name, self._outputs, data))
+
+  @staticmethod
+  def _build_sql(name, outputs, data):
+    """Creates a BigQuery SQL UDF query invocation object.
+
+    Args:
+      name: the name of the UDF function
+      outputs: a list of (name, type) tuples representing the schema of output.
+      data: the query or table over which the UDF operates.
+    """
+    return '(SELECT %s FROM %s(%s))' % (', '.join([f[0] for f in outputs]), name, data._repr_sql_())
+
+  @staticmethod
+  def _build_js(inputs, outputs, name, implementation, support_code):
+    """Creates a BigQuery SQL UDF javascript object.
+
+    Args:
+      inputs: a list of (name, type) tuples representing the schema of input.
+      outputs: a list of (name, type) tuples representing the schema of the output.
+      name: the name of the function
+      implementation: a javascript function defining the UDF logic.
+      support_code: additional javascript code that the function can use.
+    """
+    # Construct a comma-separated list of input field names
+    # For example, field1,field2,...
+    input_fields = json.dumps([f[0] for f in inputs])
+
+    # Construct a json representation of the output schema
+    # For example, [{'name':'field1','type':'string'},...]
+    output_fields = [{'name': f[0], 'type': f[1]} for f in outputs]
+    output_fields = json.dumps(output_fields, sort_keys=True)
+
+    # Build the JS from the individual bits with proper escaping of the implementation
+    if support_code is None:
+      support_code = ''
+    return '%s\n%s=%s;\nbigquery.defineFunction(\'%s\', %s, %s, %s);' % \
```

Comment:
I think we should shift to named args... if not in this change, and sometime soon. These quickly become unreadable.